server/src/main/java/org/elasticsearch/node/ResponseCollectorService.java

```diff
@@ -161,12 +164,12 @@ private double innerRank(long outstandingRequests) {
 
             // EWMA of response time
             double rS = responseTime / FACTOR;
-            // EWMA of service time
-            double muBarS = serviceTime / FACTOR;
+            // EWMA of service time. We match the paper's notation, which
+            // defines service time as the inverse of service rate (muBarS).
+            double muBarSInverse = serviceTime / FACTOR;
 
             // The final formula
-            double rank = rS - (1.0 / muBarS) + (Math.pow(qHatS, queueAdjustmentFactor) / muBarS);
-            return rank;
+            return rS - muBarSInverse + Math.pow(qHatS, queueAdjustmentFactor) * muBarSInverse;
```

Comment:
I also found this surprising and don't understand the reasoning. I am guessing it is meant as a loose approximation to the number of clients, since by default every node can serve as a coordinating node.

I can see a couple options. We could avoid making changes to `clientNum` in this PR to keep it well-scoped, while recognizing that it may give too much weight to the queue size factor. Or we could always set `clientNum` to 1 for now, which can underestimate the queue size, but is simpler and makes the calculation more predictable.

In any case, I will make sure we track this through an issue or other means.

Reply:
My worry is primarily that if we split changes here, we end up with some customers seeing multiple changes to the behavior over releases.

On the other hand, my intuition on this matter is not really strong enough that I think it is a show-stopper to merge this first and then deal with the other part later. It seems likely enough that `clientNum` will stabilize around a similar number over time for "client" nodes in the cluster. It might be too high (or low) though and building up the `numClients` could take time after restarts.

The dedicated role for coordinating nodes might come in handy here.

Reply:
This change identified a few aspects of ARS that could use improvement (including the 'stats adjustment` @jimczi mentioned above). So there will likely be more changes even apart from `clientNum`. To me it seems okay to introduce fixes/ improvements incrementally instead of assembling a single large update to the algorithm.